### PR TITLE
Desugar csend nodes specially when called from extract to variable

### PR DESCRIPTION
**Pull request by neilparikh**
Review commit-by-commit.

`a&.foo` will desugar to `a.desugar_csend_foo`, where `desugar_csend_foo` is a special name kind to signify this actually represents a `CSend`.

### Motivation

We desugar `a&.foo` to `begin; temp = a; if nil === temp then nil else temp.foo end end` currently. This causes problems for extract to variable, since we'll try to insert inside the fake `InsSeq`. I had a workaround in #8130, but that doesn't handle all cases.

Instead, we can just not do this when `desugar` is called from extract to variable. We can desugar it to regular send instead, but with a special name so that it's treated different from an actual send to that method.

### Test plan

See included automated tests.

### CODE REVIEW

File: ast/desugar/Desugar.cc [hunk 1]
```diff
@@ -84,10 +84,11 @@ ExpressionPtr numparamTree(DesugarContext dctx, int num, parser::NodeVec *decls)
     return MK::Local(core::LocOffsets::none(), name);
 }
 
-ExpressionPtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what);
+ExpressionPtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what, bool calledFromExtractToVariable);
```
Comment (jez) [inline ast/desugar/Desugar.cc:87]: We should put this in the `DesugarContext` which is already threaded to all of these calls.

**Comment by neilparikh** [comment]
Ran sorbet over Stripe's codebase, and the run time was roughly the same before and after:

Before:
```
(ins)$ pay exec rm -rf /tmp/sorbet-cache && pay exec time scripts/bin/typecheck
No errors! Great job.
556.61user 26.27system 0:54.09elapsed 1077%CPU (0avgtext+0avgdata 12730440maxresident)k
810960inputs+24outputs (13major+4696660minor)pagefaults 0swaps
(ins)$ pay exec rm -rf /tmp/sorbet-cache && pay exec time scripts/bin/typecheck
No errors! Great job.
558.31user 23.99system 0:53.95elapsed 1079%CPU (0avgtext+0avgdata 12782880maxresident)k
9120inputs+16outputs (7major+4683969minor)pagefaults 0swaps
```

After:
```
(ins)$ pay exec rm -rf /tmp/sorbet-cache && pay exec time scripts/bin/typecheck
No errors! Great job.
559.52user 24.81system 0:53.65elapsed 1089%CPU (0avgtext+0avgdata 12680704maxresident)k
1840inputs+24outputs (4major+4773410minor)pagefaults 0swaps
(ins)$ pay exec rm -rf /tmp/sorbet-cache && pay exec time scripts/bin/typecheck
No errors! Great job.
558.53user 24.89system 0:53.89elapsed 1082%CPU (0avgtext+0avgdata 12747812maxresident)k
16648inputs+16outputs (0major+4767171minor)pagefaults 0swaps
```

File: ast/desugar/Desugar.cc [hunk 2]
```diff
@@ -26,13 +26,14 @@ struct DesugarContext final {
     core::NameRef enclosingMethodName;
     bool inAnyBlock;
     bool inModule;
+    bool calledFromExtractToVariable;
```
Comment (jez) [inline ast/desugar/Desugar.cc:29]: I'm generally -1 on naming things "calledFromX"

The option should describe what behavior is being done. So like "preserveConcreteSyntax" or something like that.

File: ast/desugar/Desugar.h [hunk 3]
```diff
@@ -7,7 +7,8 @@
 
 namespace sorbet::ast::desugar {
 
-ExpressionPtr node2Tree(core::MutableContext ctx, std::unique_ptr<parser::Node> what);
+ExpressionPtr node2Tree(core::MutableContext ctx, std::unique_ptr<parser::Node> what,
+                        bool calledFromExtractToVariable = false);
```
Comment (jez) [inline ast/desugar/Desugar.h:11]: when we rename the variable let's be sure to change it here too

File: core/Names.cc [hunk 4]
```diff
@@ -78,6 +78,9 @@ string NameRef::showRaw(const GlobalState &gs) const {
                 case UniqueNameKind::Packager:
                     kind = "G";
                     break;
+                case UniqueNameKind::DesugarCsend:
+                    kind = "&";
```
Comment (jez) [inline core/Names.cc:82]: neat trick, I hadn't though to use punctuation in this.

File: main/lsp/LSPTypechecker.cc [hunk 5]
```diff
@@ -687,7 +687,7 @@ std::vector<std::unique_ptr<core::Error>> LSPTypechecker::retypecheck(vector<cor
 }
 
 ast::ExpressionPtr LSPTypechecker::getLocalVarTrees(core::FileRef fref) const {
-    auto afterDesugar = pipeline::desugarOne(config->opts, *gs, fref);
+    auto afterDesugar = pipeline::desugarOne(config->opts, *gs, fref, true);
```
Comment (jez) [inline main/lsp/LSPTypechecker.cc:690]: Preference for naming or commenting this `true` literal.

**Review by jez — COMMENTED** [review]
mostly looks good, my comments are mostly around naming / clarity.

**Review by jez — APPROVED** [review]

**Pull request merged**